src/main/java/bio/terra/workspace/service/resource/controlled/flight/delete/DeleteControlledResourceGcsBucketFlight.java

```diff
@@ -0,0 +1,174 @@
+  // for three days.
+  static class DeleteGcsBucketStep implements Step {
+    private static final int MAX_DELETE_TRIES = 72; // 3 days
+    private final CrlService crlService;
+    private final ResourceDao resourceDao;
+    private final WorkspaceService workspaceService;
+
+    public DeleteGcsBucketStep(
+        CrlService crlService, ResourceDao resourceDao, WorkspaceService workspaceService) {
+      this.crlService = crlService;
+      this.resourceDao = resourceDao;
+      this.workspaceService = workspaceService;
+    }
+
+    @Override
+    public StepResult doStep(FlightContext flightContext) throws InterruptedException {
+      int deleteTries = 0;
+
+      FlightMap inputParameters = flightContext.getInputParameters();
+      UUID workspaceId =
+          UUID.fromString(inputParameters.get(WorkspaceFlightMapKeys.WORKSPACE_ID, String.class));
+      UUID resourceId =
+          UUID.fromString(
+              inputParameters.get(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_ID, String.class));
+
+      String projectId = workspaceService.getGcpProject(workspaceId);
+
+      WsmResource wsmResource = resourceDao.getResource(workspaceId, resourceId);
+      ControlledGcsBucketResource resource =
+          wsmResource.castControlledResource().castToGcsBucketResource();
+
+      final StorageCow storageCow = crlService.createStorageCow(projectId);
+
+      BucketCow bucket = storageCow.get(resource.getBucketName());
+
+      boolean bucketExists = true;
+      while (bucketExists) {
+        // We always replace the lifecycle rules. This covers the case where the step is rerun
+        // and covers the case where the rules are changed out of band of this operation.
+        BucketCow bucketCow =
+            bucket.toBuilder()
+                .setLifecycleRules(
+                    ImmutableList.of(
+                        new BucketInfo.LifecycleRule(
+                            BucketInfo.LifecycleRule.LifecycleAction.newDeleteAction(),
+                            BucketInfo.LifecycleRule.LifecycleCondition.newBuilder()
+                                .setAge(0)
+                                .build())))
+                .build();
+        bucketCow.update();
+        bucketExists = tryBucketDelete(bucket);
+        if (bucketExists) {
+          TimeUnit.HOURS.sleep(1);
+        }
+        deleteTries++;
+        if (deleteTries >= MAX_DELETE_TRIES) {
+          // This will cause the flight to fail.
+          throw new BucketDeleteTimeoutException(
+              String.format("Failed to delete bucket after %d hours", MAX_DELETE_TRIES));
+        }
+      }
+      return StepResult.getStepResultSuccess();
+    }
+
+    /**
+     * Try deleting the bucket. It will fail if there are objects still in the bucket.
+     *
+     * @param bucket bucket we should try to delete
+     * @return bucket existence: true if the bucket still exists; false if we deleted it
```

Comment:
I think this is the opposite sense of a delete method or two I either wrote or saw somewhere. It was true for deleted, false for not deleted (still exists).

Reply:
Yes, but in this case, it is the right sense for the while loop and if. Otherwise, I get the complaint that all uses of the return value are negated.